File: leetcode/2552.统计上升四元组.cpp
int v[4000];
class Solution {
public:
    long long countQuadruplets(vector<int> &nums) {
        ios::sync_with_stdio(false);
        cin.tie(nullptr);
        int n=nums.size();
        long long sum=0;
        memset(v,0,sizeof(v));
        for(int i=2;i<n;i++){
            int z=0;
            for(int j=0;j<i;j++){
                if(nums[i]>nums[j]){
                    sum+=v[j];
                    z++;
                }else{
                    v[j]+=z;
                }
            }
        }
        return sum;
    }
};

File: leetcode/2552.统计上升四元组.py
class Solution:
    def countQuadruplets(self, nums: List[int]) -> int:
        n=len(nums)
        great=[0]*n
        great[-1]=[0]*(n+1)
        for k in range(n-2,-1,-1):
            great[k]=great[k+1][::]
            for x in range(1,nums[k+1]):
                great[k][x]+=1
        less=[0]*n
        less[0]=[0]*(n+1)
        for j in range(1,n-1):
            less[j]=less[j-1][::]
            for x in range(nums[j-1]+1,n+1):
                less[j][x]+=1
        # 枚举j和k
        ans=0
        for j in range(1,n-2):
            for k in range(j+1,n-1):
                if nums[j]>nums[k]:
                    ans+=less[j][nums[k]]*great[k][nums[j]]
        return ans